ts/src/dna_factory_v2.ts

```diff
@@ -0,0 +1,313 @@
+  private getLatestSchemaSubversion(schemaVersion?: string): string {
+    if (!schemaVersion) return LATEST_SCHEMA_VERSION;
+    else if (schemaVersion?.includes('.')) return schemaVersion;
+    else if (schemaVersion && this.latestSchemasSubversions[schemaVersion])
+      return this.latestSchemasSubversions[schemaVersion];
+    const completeVersion = getLatestSubversion(dnaSchemas, schemaVersion);
+    this.latestSchemasSubversions[schemaVersion] = completeVersion;
+    return completeVersion;
+  }
+
+  private _getRandomRarity(grade: Grade): Rarity {
+    const rarities = raritiesJson[grade];
+    if (!rarities) {
+      throw new Error(`No rarity found for input ${grade}`);
+    }
+    const precision = 3;
+    const multiplier = Math.pow(10, precision);
+    const weightsSum = Object.values(rarities).reduce((acc, rarity) => acc + rarity.probability * multiplier, 0);
+    const random = Math.random() * weightsSum;
+    let total = 0;
+    for (const [rarity, rarityInfo] of Object.entries(rarities)) {
+      total += rarityInfo.probability * multiplier;
+      if (random <= total) return rarity as Rarity;
+    }
+    throw new Error(`No rarity found: ${weightsSum}, ${random}`);
+  }
+
+  getDNASchema(version?: version): DNASchemaV4 {
+    if (!version) return dnaSchemas[LATEST_SCHEMA_VERSION];
+    else if (dnaSchemas[version]) return dnaSchemas[version];
+    else if (this.latestSchemasSubversions[version]) return dnaSchemas[this.latestSchemasSubversions[version]];
+
+    const completeVersion = version.includes('.') ? version : this.getLatestSchemaSubversion(version);
+
+    if (!completeVersion) throw new Error(`No schema found for ${version}`);
+
+    const dnaSchema: DNASchemaV4 = dnaSchemas[completeVersion];
+    if (completeVersion !== dnaSchema.version)
+      throw new Error(`Versions mismatch: ${completeVersion} (filename) vs ${dnaSchema.version} (schema)`);
+    return dnaSchemas[completeVersion];
+  }
+
+  private serializeDna(data: DnaDataV2): string {
+    return compressToBase64(JSON.stringify(data));
+  }
+
+  private deserializeDna(dna: string): DnaDataV2 {
+    return JSON.parse(decompressFromBase64(dna)) as DnaDataV2;
+  }
+
+  private getDna(version: version, data: DnaDataV2): string {
+    const versionDNAFormat = toPaddedHexa(version, 4);
+    const serializedData = this.serializeDna(data);
+    const dna = versionDNAFormat + serializedData;
+    return dna;
+  }
+
+  /**
+   * Generate statsCount number of stats with a mean value in the rarity range.
+   * @param rarity Rarity
+   * @param statsCount Number of stats to generate
+   */
+  private _generateStatsForRarity(nStats: number, grade: Grade, rarity: Rarity): number[] {
+    const [minStatAvg, maxStatAvg] = raritiesJson[grade][rarity].average_stats_range;
+    const stats = Array.from(Array(nStats).keys()).map(() => 0);
+
+    const mean = randomInt(minStatAvg, maxStatAvg, true);
+    // adding up to 5 will still result in the same mean as we are rounding down
```

Comment:
What does this mean?

Reply:
I moved it over if (Math.floor(average) !== mean) pointsLeft += 1; to justify why we can add points left and still get the same average